### ISSUE TITLE
polars can no longer find column in LazyFrame

### ISSUE THREAD

**Issue by harrymconner** [pola-rs/polars#16505]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

The below code works on 0.20.26, but not on 0.20.28 or 0.20.29. I did not try 0.20.27 because I saw it was yanked. The code is just doing some financial transaction processing and adds some columns that are needed in subsequent steps of the pipeline. I'm getting the following error in the next step of the pipeline: polars.exceptions.ColumnNotFoundError: IRRGLDate. This is strange because I'm adding the IRRGLDate column in irr_gl_date() (see the final with_columns() call).
```python
def prepare_for_irr(transactions: pl.LazyFrame) -> pl.LazyFrame:
    return (
        transactions.with_columns(
            first_date_col(),
            has_beg_col(),
            last_date_col(),
            has_end_col(),
        )
        .with_columns(add_short_col())
        .with_columns(short_nav_adj())
        .with_columns(irr_gl_date(), irr_amount(), short_adj_irr_amount())  # IRRGLDate column added here in irr_gl_date()
    )
```

Interestingly, while I was debugging, I added an inspect statement before the final with_columns call and doing so caused the pipeline to succeed on 0.20.29. Somehow the inspect is helping polars "find" the IRRGLDate column?
```python
def prepare_for_irr(transactions: pl.LazyFrame) -> pl.LazyFrame:
    return (
        transactions.with_columns(
            first_date_col(),
            has_beg_col(),
            last_date_col(),
            has_end_col(),
        )
        .with_columns(add_short_col())
        .with_columns(short_nav_adj())
        .inspect()  # adding the inspect here fixes the issue
        .with_columns(irr_gl_date(), irr_amount(), short_adj_irr_amount())
    )
```

### Log output

```shell
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
found multiple sources; run comm_subplan_elim
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
CACHE SET: cache id: 0
CACHE HIT: cache id: 0
dataframe filtered
SEMI join dataframes finished
found multiple sources; run comm_subplan_elim
join parallel: false
UNION: `parallel=false` union is run sequentially
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
dataframe filtered
INNER join dataframes finished
CACHE SET: cache id: 0
CACHE HIT: cache id: 0
dataframe filtered
SEMI join dataframes finished
found multiple sources; run comm_subplan_elim
join parallel: false
UNION: `parallel=false` union is run sequentially
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
dataframe filtered
INNER join dataframes finished
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
UNION: `parallel=false` union is run sequentially
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
dataframe filtered
INNER join dataframes finished
UNION: `parallel=false` union is run sequentially
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
dataframe filtered
INNER join dataframes finished
dataframe filtered
INNER join dataframes finished
CACHE SET: cache id: 0
CACHE HIT: cache id: 0
dataframe filtered
SEMI join dataframes finished
found multiple sources; run comm_subplan_elim
UNION: `parallel=false` union is run sequentially
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
INNER join dataframes finished
UNION: `parallel=false` union is run sequentially
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
dataframe filtered
INNER join dataframes finished
CACHE SET: cache id: 1
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
CACHE HIT: cache id: 1
CACHE HIT: cache id: 1
dataframe filtered
INNER join dataframes finished
dataframe filtered
dataframe filtered
read files in parallel
avg line length: 76.49902
std. dev. line length: 12.226215
initial row estimate: 58407
no. of chunks: 16 processed by: 16 threads.
found multiple sources; run comm_subplan_elim
join parallel: false
UNION: `parallel=false` union is run sequentially
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 76.49902
std. dev. line length: 12.226215
initial row estimate: 58407
no. of chunks: 16 processed by: 16 threads.
read files in parallel
avg line length: 76.49902
std. dev. line length: 12.226215
initial row estimate: 58407
no. of chunks: 16 processed by: 16 threads.
LEFT join dataframes finished
dataframe filtered
CACHE SET: cache id: 0
dataframe filtered
CACHE HIT: cache id: 0
dataframe filtered
UNION: `parallel=false` union is run sequentially
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
INNER join dataframes finished
UNION: `parallel=false` union is run sequentially
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
dataframe filtered
INNER join dataframes finished
CACHE SET: cache id: 2
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
CACHE HIT: cache id: 2
CACHE HIT: cache id: 2
dataframe filtered
INNER join dataframes finished
INNER join dataframes finished
dataframe filtered
dataframe filtered
found multiple sources; run comm_subplan_elim
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 114.74805
std. dev. line length: 15.587312
initial row estimate: 149085
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
UNION: `parallel=false` union is run sequentially
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
INNER join dataframes finished
UNION: `parallel=false` union is run sequentially
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
keys/aggregates are not partitionable: running default HASH AGGREGATION
read files in parallel
avg line length: 55.066406
std. dev. line length: 13.627883
initial row estimate: 115796
no. of chunks: 16 processed by: 16 threads.
dataframe filtered
dataframe filtered
dataframe filtered
INNER join dataframes finished
CACHE SET: cache id: 1
keys/aggregates are not partitionable: running default HASH AGGREGATION
join parallel: false
CACHE HIT: cache id: 1
CACHE HIT: cache id: 1
dataframe filtered
INNER join dataframes finished
INNER join dataframes finished
found multiple sources; run comm_subplan_elim
```

### Issue description

I have an ETL pipeline that had been working perfectly on polars 0.20.26, but the pipeline is now failing after upgrading to 0.20.29. The pipeline can no longer "find" a column that is added in the prior step. I tried my best to create a useful MRE but since I can't share the underlying data or any other code from the pipeline, I have been unsuccessful so far. I tried to provide as much information as possible in the MRE section, but please let me know if any additional information would be helpful!

### Expected behavior

I expect the pipeline to be able to find the IRRGLDate column in the next step of the pipeline.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.29
Index type:           UInt32
Platform:             Windows-10-10.0.22631-SP0
Python:               3.11.6 (tags/v3.11.6:8b6ee5b, Oct  2 2023, 14:57:12) [MSC v.1935 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           0.3.3
deltalake:            <not installed>
fastexcel:            0.10.4
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           <not installed>
nest_asyncio:         <not installed>
numpy:                1.26.4
openpyxl:             <not installed>
pandas:               2.2.2
pyarrow:              16.0.0
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           3.2.0
```

</details>

**Comment by cmdlineluser** [comment]
This is probably the same ColumnNotFound issue that is fixed on main https://github.com/pola-rs/polars/issues/16435 (there is a new cluster_with_columns optimization)

I believe `0.20.30` is about to be released which includes it.

**Comment by harrymconner** [comment]
Confirming this is fixed on 0.20.30! Impressive turnaround time!
